go/updater/trainingJobUpdater.go

```diff
@@ -48,34 +53,417 @@ type TrainingJobUpdater struct {
+				log.Infof("Release master, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Trainer.ReplicaSpec.Name)
+				if err := updater.releaseMaster(); err != nil {
+					log.Error(err.Error())
+				}
+			}
+			log.Infof("Release pserver, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Trainer.ReplicaSpec.Name)
+			if err := updater.releasePserver(); err != nil {
+				log.Error(err.Error())
+			}
+		}
+	}
+
+	if updater.status.Phase == padv1.TrainingJobPhaseSucceeded || updater.status.Phase == padv1.TrainingJobPhaseFailed {
+		if ticker != nil {
+			log.Infof("stop ticker, namespace=%v name=%v: ", updater.job.Namespace, updater.job.Name)
+			ticker.Stop()
+		}
+	}
+	return
+}
+
+// InitResource is used to parse trainingJob and create trainingJob resources.
+func (updater *TrainingJobUpdater) InitResource() {
+	if updater.status.Phase == padv1.TrainingJobPhaseNone {
+		log.Infof("set up trainingJob namespace=%v name=%v: ", updater.job.Namespace, updater.job.Name)
+		updater.parseTrainingJob()
+		err := updater.updateCRDStatus()
+		if err != nil {
+			log.Warning("set up trainingJob to update trainingJob status error: ", err.Error())
+		}
 	}
 
-	go jobber.start()
-	return jobber, nil
+	if updater.status.Phase == padv1.TrainingJobPhaseCreating {
+		log.Infof("create trainingJob namespace=%v name=%v: ", updater.job.Namespace, updater.job.Name)
+		_ = updater.createTrainingJob()
+		err := updater.updateCRDStatus()
+		if err != nil {
+			log.Warning("create trainingJob to update trainingJob status error: ", err.Error())
+		}
+		if updater.status.Phase == padv1.TrainingJobPhaseFailed {
+			log.Infof("Release Resource for failed namespace=%v name=%v: ", updater.job.Namespace, updater.job.Name)
+			if updater.job.Spec.FaultTolerant {
+				log.Infof("Release master, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Trainer.ReplicaSpec.Name)
+				if err := updater.releaseMaster(); err != nil {
+					log.Error(err.Error())
+				}
+			}
+
+			log.Infof("Release pserver, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Trainer.ReplicaSpec.Name)
+			if err := updater.releasePserver(); err != nil {
+				log.Error(err.Error())
+			}
+		}
+	}
 }
 
 // Start is the main process of life cycle of a TrainingJob, including create resources, event process handle and
 // status convert.
 func (updater *TrainingJobUpdater) start() {
-	//	TODO(zhengqi): this will commit in the next pr
+	log.Infof("start updater, namespace=%v name=%v: ", updater.job.Namespace, updater.job.Name)
+	go updater.InitResource()
 
+	ticker := time.NewTicker(convertedTimerTicker)
+	defer ticker.Stop()
+	log.Infof("start ticker, namespace=%v name=%v: ", updater.job.Namespace, updater.job.Name)
+	for {
+		select {
+		case ev := <-updater.eventCh:
+			switch ev.pet {
+			case trainingJobEventDelete:
+				log.Infof("Delete updater, namespace=%v name=%v: ", updater.job.Namespace, updater.job.Name)
+				ticker.Stop()
```

Comment:
There is already `defer ticker.Stop()` above, maybe it's not necessary?

Reply:
Done